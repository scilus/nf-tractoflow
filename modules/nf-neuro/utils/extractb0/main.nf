process UTILS_EXTRACTB0 {
    tag "$meta.id"
    label 'process_single'

    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'https://scil.usherbrooke.ca/containers/scilus_2.0.2.sif':
        'scilus/scilus:2.0.2' }"

    input:
    tuple val(meta), path(dwi), path(bval), path(bvec)

    output:
    tuple val(meta), path("*_b0*.nii.gz"), emit: b0
    path "versions.yml"                  , emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    def prefix = task.ext.prefix ?: "${meta.id}"
    def extraction_strategy = task.ext.b0_extraction_strategy ? "--$task.ext.b0_extraction_strategy" : "--mean"
    def b0_threshold = task.ext.b0_threshold ? "--b0_threshold $task.ext.b0_threshold" : ""
    def output_series = task.ext.output_series ? "" : "--single-image"
    """
    export ITK_GLOBAL_DEFAULT_NUMBER_OF_THREADS=1
    export OMP_NUM_THREADS=1
    export OPENBLAS_NUM_THREADS=1

    scil_dwi_extract_b0.py $dwi $bval $bvec ${prefix}_b0.nii.gz \
        $output_series $extraction_strategy $b0_threshold --skip_b0_check

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.1
    END_VERSIONS
    """

    stub:
    def args = task.ext.args ?: ''
    def prefix = task.ext.prefix ?: "${meta.id}"

    """
    scil_dwi_extract_b0.py - h

    touch ${prefix}_b0.nii.gz

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        scilpy: 2.0.1
    END_VERSIONS
    """
}
