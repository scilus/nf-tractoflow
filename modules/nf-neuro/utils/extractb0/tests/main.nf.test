nextflow_process {

    name "Test Process UTILS_EXTRACTB0"
    script "../main.nf"
    process "UTILS_EXTRACTB0"

    tag "modules"
    tag "modules_nfcore"
    tag "utils"
    tag "utils/extractb0"

    tag "subworkflows"
    tag "subworkflows/load_test_data"

    setup {
            run("LOAD_TEST_DATA", alias: "LOAD_DATA") {
                script "../../../../../subworkflows/nf-neuro/load_test_data/main.nf"
                process {
                    """
                    input[0] = Channel.from( [ "topup_eddy_light.zip" ] )
                    input[1] = "test.load-test-data"
                    """
                }
            }
        }

    test("utils - extractb0") {
        config "./nextflow.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true)
                    ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("utils - extractb0 - all4d") {
        config "./nextflow_all4d.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true)
                    ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("utils - extractb0 - allseries") {
        config "./nextflow_allseries.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true)
                    ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("utils - extractb0 - clusterfirst") {
        config "./nextflow_clusterfirst.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true)
                    ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("utils - extractb0 - clustermean") {
        config "./nextflow_clustermean.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true)
                    ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("utils - extractb0 - mean") {
        config "./nextflow_mean.config"
        when {
            process {
                """
                input[0] = LOAD_DATA.out.test_data_directory
                    .map{ test_data_directory -> [
                        [ id:'test', single_end:false ], // meta map
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.nii.gz", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bval", checkIfExists: true),
                        file("\${test_data_directory}/sub-01_dir-AP_dwi.bvec", checkIfExists: true)
                    ]}
                """
            }
        }
        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }
}
