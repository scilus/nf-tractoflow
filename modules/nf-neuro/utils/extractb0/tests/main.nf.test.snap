{
    "utils - extractb0 - mean": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:14:29.554229"
    },
    "utils - extractb0 - all4d": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:13:58.708201"
    },
    "utils - extractb0": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "1": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,91288914a80e7866ef8b08a363d617f9"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:13:51.838198"
    },
    "utils - extractb0 - clustermean": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,feb9010e5b72fc6f0c814f9dd54cb710"
                    ]
                ],
                "1": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,feb9010e5b72fc6f0c814f9dd54cb710"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:14:22.350869"
    },
    "utils - extractb0 - clusterfirst": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,27e4aa847d60b2154d8865d19ba3e06e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,27e4aa847d60b2154d8865d19ba3e06e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:14:14.511257"
    },
    "utils - extractb0 - allseries": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ],
                "b0": [
                    [
                        {
                            "id": "test",
                            "single_end": false
                        },
                        "test_b0.nii.gz:md5,2066111f2b964735a168d6c653ba1a8b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,36a507648bd7506126303551eff1e6d0"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-09T00:14:07.253309"
    }
}